File: internal/templates/files/python/mxreq_docstring.py
"""
mxreq YAML Docstring Generator.
Copy this file into your test repository and import it.

Generates YAML-formatted docstrings and inserts them directly into source files.
Use the DS=1 environment variable to enable docstring generation mode.

Usage:
    # Generate docstrings (test code is skipped):
    DS=1 python -m pytest tests/test_example.py -s

    # Normal test execution (ds_* calls are no-ops):
    python -m pytest tests/test_example.py

Example:
    from mxreq_docstring import ds, ds_header, ds_step, ds_manual, ds_user, ds_footer
    from mxreq_results import verify_equal, start_test, end_test

    def test_NEW_short_test_name():
        ds_header(
            title="Long Test Name",
            description="Test description",
            folder="F-TC-123",
            assumptions=["* Assumption 1", "* Assumption 2"]
        )

        start_test()

        # Step 1: Setup
        ds_step("Description of step")
        if not ds:
            run = setup("arg1", "arg2")
            execute(run)

        # Step 2: User intervention required
        ds_user("User will perform an action", expected="Action has been performed", req="SOFT-123")

        # Step 3: Verify (automated)
        ds_step("Verify action is as per TSPEC-82", expected="Action is as per TSPEC", req="SOFT-124")
        if not ds:
            verify_equal("SOFT-124", action(), "argToCompare")

        # Step 4: Manual verification
        ds_manual("Visually inspect the display for correct formatting", expected="Time is displayed in MM:SS format", req="SOFT-125")

        # Step 5: Cleanup
        ds_step("Discard process")
        if not ds:
            discard()

        ds_footer(labels=["Automated", "VM"])
        if not ds: end_test()
"""

import os
import re
import inspect
from typing import Optional
from threading import Lock


# Environment toggle - True when DS=1 is set
ds = os.getenv("DS") == "1"


class _DocstringBuilder:
    """Thread-safe singleton builder for YAML docstrings."""

    def __init__(self):
        self._lock = Lock()
        self._current: Optional[dict] = None

    def start(
        self,
        title: str,
        description: str = "",
        folder: str = "",
        assumptions: Optional[list[str]] = None,
    ):
        """Start building a new docstring."""
        with self._lock:
            self._current = {
                "title": title,
                "description": description,
                "folder": folder,
                "assumptions": assumptions or [],
                "steps": [],
                "labels": [],
                "up_links": set(),
                "source_file": None,
                "function_name": None,
            }
            # Auto-detect source file and function from caller
            # Go up 2 frames: start() -> ds_header() -> test function
            frame = inspect.currentframe()
            if frame and frame.f_back and frame.f_back.f_back:
                caller = frame.f_back.f_back
                self._current["source_file"] = caller.f_code.co_filename
                self._current["function_name"] = caller.f_code.co_name

    def add_step(self, action: str, expected: str = "N/A", req: str = ""):
        """Add a test step to the current docstring."""
        with self._lock:
            if not self._current:
                raise RuntimeError("Call ds_header() before ds_step()")
            self._current["steps"].append(
                {
                    "action": action,
                    "expected": expected,
                    "RequirementLink": req,
                }
            )
            # Collect requirements for up_links
            if req:
                for r in req.split(","):
                    self._current["up_links"].add(r.strip())

    def finish(self, labels: Optional[list[str]] = None):
        """Finish building and insert the docstring into the source file."""
        with self._lock:
            if not self._current:
                raise RuntimeError("Call ds_header() before ds_footer()")
            if labels:
                self._current["labels"] = labels

            yaml_str = self._generate_yaml()
            self._insert_docstring(yaml_str)
            self._current = None

    def _generate_yaml(self) -> str:
        """Generate YAML content for the docstring."""
        c = self._current
        lines = ['"""', "---"]

        # Title
        lines.append(f'title: "{c["title"]}"')

        # Folder
        if c["folder"]:
            lines.append(f"folder: {c['folder']}")

        # Description (literal block scalar)
        if c["description"]:
            lines.append("description: |")
            for line in c["description"].split("\n"):
                lines.append(f"  {line}")

        # Assumptions (formatted as HTML list)
        if c["assumptions"]:
            lines.append("assumptions: |")
            lines.append("  <ul>")
            for assumption in c["assumptions"]:
                # Strip leading "* " if present and wrap in <li> tags
                text = assumption[2:] if assumption.startswith("* ") else assumption
                lines.append(f"   <li>{text}</li>")
            lines.append("  </ul>")

        # Steps
        lines.append("steps:")
        for step in c["steps"]:
            lines.append(f'  - action: "{step["action"]}"')
            if step["expected"] != "N/A":
                lines.append(f'    expected: "{step["expected"]}"')
            if step["RequirementLink"]:
                lines.append(f'    RequirementLink: "{step["RequirementLink"]}"')

        # Labels
        if c["labels"]:
            lines.append("labels:")
            for label in c["labels"]:
                lines.append(f"  - {label}")

        # Up links (collected from steps)
        if c["up_links"]:
            links = ", ".join(sorted(c["up_links"]))
            lines.append(f'up_links: "{links}"')

        lines.append("---")
        lines.append('"""')

        return "\n".join(lines)

    def _insert_docstring(self, yaml_str: str):
        """Insert or replace the docstring in the source file."""
        source_file = self._current["source_file"]
        func_name = self._current["function_name"]

        with open(source_file, "r") as f:
            content = f.read()

        # Pattern to find function definition and optional existing docstring
        # Matches: def func_name(...): followed by optional docstring
        pattern = rf'(def {re.escape(func_name)}\([^)]*\):)\s*\n(\s*"""[\s\S]*?"""\s*\n)?'

        def replacer(match):
            func_def = match.group(1)
            # Indent the YAML docstring with 4 spaces
            indent = "    "
            indented_yaml = "\n".join(indent + line for line in yaml_str.split("\n"))
            return f"{func_def}\n{indented_yaml}\n"

        new_content, count = re.subn(pattern, replacer, content)

        if count == 0:
            print(f"Could not find function '{func_name}' in {source_file}")
            return

        with open(source_file, "w") as f:
            f.write(new_content)

        print(f"Generated docstring for {func_name} in {source_file}")


# Global singleton
_builder = _DocstringBuilder()


# Public API
def ds_header(
    title: str,
    description: str = "",
    folder: str = "",
    assumptions: Optional[list[str]] = None,
):
    """Start building a test case docstring.

    Args:
        title: Test case title as it will appear in Matrix.
        description: HTML-formatted description of the test.
        folder: Parent folder reference (e.g., "F-TC-521").
        assumptions: List of assumption/precondition strings.
    """
    if ds:
        _builder.start(title, description, folder, assumptions)


def ds_step(action: str, expected: str = "N/A", req: str = ""):
    """Add a test step to the docstring.

    Args:
        action: Description of the test action.
        expected: Expected result. Defaults to "N/A" for setup/cleanup steps.
        req: Requirement link(s), comma-separated (e.g., "SOFT-387" or "SOFT-387,SOFT-388").
    """
    if ds:
        _builder.add_step(action, expected, req)


def ds_manual(action: str, expected: str = "N/A", req: str = ""):
    """Convenience wrapper around ds_step that prepends '<b>Manual:</b> ' to the action.

    Equivalent to: ds_step(f"<b>Manual:</b> {action}", expected, req)

    Args:
        action: Description of the test action.
        expected: Expected result. Defaults to "N/A" for setup/cleanup steps.
        req: Requirement link(s), comma-separated (e.g., "SOFT-387" or "SOFT-387,SOFT-388").
    """
    ds_step(f"<b>Manual:</b> {action}", expected, req)


def ds_user(action: str, expected: str = "N/A", req: str = ""):
    """Convenience wrapper around ds_step that prepends '<b><i>USER INTERVENTION</b></i> ' to the action.

    Equivalent to: ds_step(f"<b><i>USER INTERVENTION</b></i> {action}", expected, req)

    Args:
        action: Description of the test action.
        expected: Expected result. Defaults to "N/A" for setup/cleanup steps.
        req: Requirement link(s), comma-separated (e.g., "SOFT-387" or "SOFT-387,SOFT-388").
    """
    ds_step(f"<b><i>USER INTERVENTION</b></i> {action}", expected, req)


def ds_footer(labels: Optional[list[str]] = None):
    """Finish building and insert the docstring into the source file.

    Args:
        labels: List of label strings (e.g., ["Automated", "VM"]).
    """
    if ds:
        _builder.finish(labels)

File: internal/templates/files/go/mxreq_docstring.go
/*
mxreq YAML Docstring Generator for Go.
Copy this file into your test repository and import it.

Generates YAML-formatted docstrings and inserts them directly into source files.
Use the DS=1 environment variable to enable docstring generation mode.

Usage:

	# Generate docstrings (test code is skipped):
	DS=1 go test ./...

	# Normal test execution (Ds* calls are no-ops):
	go test ./...

Example:

	func Test_NEW_short_test_name(t *testing.T) {
	    mxreq.DsHeader("Long Test Name",
	        "Test description",
	        "F-TC-123",
	        "* Assumption 1",
	        "* Assumption 2")

	    mxreq.StartTest(t)

	    // Step 1: Setup
	    mxreq.DsStep("Description of step", "", "")
	    if !mxreq.Ds {
	        run := setup("arg1", "arg2")
	        execute(run)
	    }

	    // Step 2: User intervention required
	    mxreq.DsUser("User will perform an action", "Action has been performed", "SOFT-123")

	    // Step 3: Verify (automated)
	    mxreq.DsStep("Verify action is as per TSPEC-82", "Action is as per TSPEC", "SOFT-124")
	    if !mxreq.Ds {
	        mxreq.VerifyEqual(t, "SOFT-124", action(), "argToCompare")
	    }

	    // Step 4: Manual verification
	    mxreq.DsManual("Visually inspect the display for correct formatting", "Time is displayed in MM:SS format", "SOFT-125")

	    // Step 5: Cleanup
	    mxreq.DsStep("Discard process", "", "")
	    if !mxreq.Ds {
	        discard()
	    }

	    mxreq.DsFooter("Automated", "VM")
	    if !mxreq.Ds { mxreq.EndTest(t) }
	}
*/
package mxreq

import (
	"fmt"
	"os"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"sync"
)

// Ds is true when DS=1 environment variable is set.
// Use this to conditionally skip test code during docstring generation.
var Ds = os.Getenv("DS") == "1"

type testStep struct {
	Action          string
	Expected        string
	RequirementLink string
}

type docstringBuilder struct {
	mu          sync.Mutex
	title       string
	description string
	folder      string
	assumptions []string
	steps       []testStep
	labels      []string
	upLinks     map[string]bool
	sourceFile  string
	funcName    string
}

var currentBuilder *docstringBuilder

// DsHeader starts building a test case docstring.
//
// Parameters:
//   - title: Test case title as it will appear in Matrix.
//   - description: Description of the test.
//   - folder: Parent folder reference (e.g., "F-TC-521").
//   - assumptions: Variadic list of assumption/precondition strings.
func DsHeader(title, description, folder string, assumptions ...string) {
	if !Ds {
		return
	}

	// Get caller info
	pc, file, _, ok := runtime.Caller(1)
	if !ok {
		fmt.Println("Could not determine caller info")
		return
	}

	funcName := runtime.FuncForPC(pc).Name()
	// Extract just the function name (remove package path)
	parts := strings.Split(funcName, ".")
	funcName = parts[len(parts)-1]

	currentBuilder = &docstringBuilder{
		title:       title,
		description: description,
		folder:      folder,
		assumptions: assumptions,
		steps:       []testStep{},
		labels:      []string{},
		upLinks:     make(map[string]bool),
		sourceFile:  file,
		funcName:    funcName,
	}
}

// DsStep adds a test step to the docstring.
//
// Parameters:
//   - action: Description of the test action.
//   - expected: Expected result. Use "" for "N/A" (setup/cleanup steps).
//   - req: Requirement link(s), comma-separated (e.g., "SOFT-387").
func DsStep(action, expected, req string) {
	if !Ds || currentBuilder == nil {
		return
	}

	currentBuilder.mu.Lock()
	defer currentBuilder.mu.Unlock()

	if expected == "" {
		expected = "N/A"
	}

	currentBuilder.steps = append(currentBuilder.steps, testStep{
		Action:          action,
		Expected:        expected,
		RequirementLink: req,
	})

	// Collect requirements for up_links
	if req != "" {
		for _, r := range strings.Split(req, ",") {
			currentBuilder.upLinks[strings.TrimSpace(r)] = true
		}
	}
}

// DsManual is a convenience wrapper around DsStep that prepends '<b>Manual:</b> ' to the action.
//
// Equivalent to: DsStep("<b>Manual:</b> "+action, expected, req)
//
// Parameters:
//   - action: Description of the test action.
//   - expected: Expected result. Use "" for "N/A" (setup/cleanup steps).
//   - req: Requirement link(s), comma-separated (e.g., "SOFT-387").
func DsManual(action, expected, req string) {
	DsStep("<b>Manual:</b> "+action, expected, req)
}

// DsUser is a convenience wrapper around DsStep that prepends '<b><i>USER INTERVENTION</b></i> ' to the action.
//
// Equivalent to: DsStep("<b><i>USER INTERVENTION</b></i> "+action, expected, req)
//
// Parameters:
//   - action: Description of the test action.
//   - expected: Expected result. Use "" for "N/A" (setup/cleanup steps).
//   - req: Requirement link(s), comma-separated (e.g., "SOFT-387").
func DsUser(action, expected, req string) {
	DsStep("<b><i>USER INTERVENTION</b></i> "+action, expected, req)
}

// DsFooter finishes building and inserts the docstring into the source file.
//
// Parameters:
//   - labels: Variadic list of label strings (e.g., "Automated", "VM").
func DsFooter(labels ...string) {
	if !Ds || currentBuilder == nil {
		return
	}

	currentBuilder.mu.Lock()
	defer currentBuilder.mu.Unlock()

	currentBuilder.labels = labels

	yaml := generateYAML()
	insertDocstring(yaml)
	currentBuilder = nil
}

func generateYAML() string {
	b := currentBuilder
	var lines []string

	lines = append(lines, "/*")
	lines = append(lines, "---")

	// Title
	lines = append(lines, fmt.Sprintf(`title: "%s"`, b.title))

	// Folder
	if b.folder != "" {
		lines = append(lines, fmt.Sprintf("folder: %s", b.folder))
	}

	// Description (literal block scalar)
	if b.description != "" {
		lines = append(lines, "description: |")
		for _, line := range strings.Split(b.description, "\n") {
			lines = append(lines, fmt.Sprintf("  %s", line))
		}
	}

	// Assumptions (formatted as HTML list)
	if len(b.assumptions) > 0 {
		lines = append(lines, "assumptions: |")
		lines = append(lines, "  <ul>")
		for _, a := range b.assumptions {
			// Strip leading "* " if present and wrap in <li> tags
			text := a
			if strings.HasPrefix(a, "* ") {
				text = a[2:]
			}
			lines = append(lines, fmt.Sprintf("   <li>%s</li>", text))
		}
		lines = append(lines, "  </ul>")
	}

	// Steps
	lines = append(lines, "steps:")
	for _, step := range b.steps {
		lines = append(lines, fmt.Sprintf(`  - action: "%s"`, step.Action))
		if step.Expected != "N/A" {
			lines = append(lines, fmt.Sprintf(`    expected: "%s"`, step.Expected))
		}
		if step.RequirementLink != "" {
			lines = append(lines, fmt.Sprintf(`    RequirementLink: "%s"`, step.RequirementLink))
		}
	}

	// Labels
	if len(b.labels) > 0 {
		lines = append(lines, "labels:")
		for _, l := range b.labels {
			lines = append(lines, fmt.Sprintf("  - %s", l))
		}
	}

	// Up links (collected from steps)
	if len(b.upLinks) > 0 {
		var links []string
		for l := range b.upLinks {
			links = append(links, l)
		}
		sort.Strings(links)
		lines = append(lines, fmt.Sprintf(`up_links: "%s"`, strings.Join(links, ", ")))
	}

	lines = append(lines, "---")
	lines = append(lines, "*/")

	return strings.Join(lines, "\n")
}

func insertDocstring(yaml string) {
	b := currentBuilder

	content, err := os.ReadFile(b.sourceFile)
	if err != nil {
		fmt.Printf("Error reading file %s: %v\n", b.sourceFile, err)
		return
	}

	// Pattern to find function definition and optional existing docstring
	// Matches: func FuncName(...) { followed by optional block comment
	funcPattern := regexp.MustCompile(
		fmt.Sprintf(`(func %s\([^)]*\)[^{]*\{)\s*\n(\s*/\*[\s\S]*?\*/\s*\n)?`,
			regexp.QuoteMeta(b.funcName)))

	// Indent the YAML with a tab (Go convention)
	indentedYAML := strings.ReplaceAll(yaml, "\n", "\n\t")
	replacement := fmt.Sprintf("$1\n\t%s\n", indentedYAML)

	newContent := funcPattern.ReplaceAllString(string(content), replacement)

	if newContent == string(content) {
		fmt.Printf("Could not find function '%s' in %s\n", b.funcName, b.sourceFile)
		return
	}

	err = os.WriteFile(b.sourceFile, []byte(newContent), 0644)
	if err != nil {
		fmt.Printf("Error writing file %s: %v\n", b.sourceFile, err)
		return
	}

	fmt.Printf("Generated docstring for %s in %s\n", b.funcName, b.sourceFile)
}
